### ISSUE TITLE
What is next?

### ISSUE THREAD

**Issue by SenexCrenshaw** [SenexCrenshaw/xTeVe#53]
I am debating what the next steps here should be. I was considering:

1. A UI rewrite with something like Vue.JS
2. The backend probably could use some work and porting to C# is interesting as well.
3. Name change??

I would separate the two systems (frontend/backend). I think these two will allow other features to be added more easily. Things like channel failover #47  and #38 

I am fine with the way things are since it works but I am always considering new ideas that might help y'all out.

**Comment by SenexCrenshaw** [comment]
I decided to create a new repo. It will use c# in the backend for the server/api piece and VUE JS3 for the front end. I will keep updating this repo as bugs come in. 

With the new application:
  Virtual hdhomeruns with individual channels.
  Multiple channels can be grouped together for additional viewing connections/failover
  EPG data can be selected per channel. 

This should give lots of flexibility and a new UI will be welcomed.

**Comment by aniel300** [comment]
yes please a revamp might not be a bad idea

**Comment by SenexCrenshaw** [comment]
things are moving along nicely. I have a decent backend written in .net core c# and the UI in react (TS). The plan is to allow uploading whatever amount of M3U/EPG files. From there you can allocate any channels from the EPG files to any amount of HD Home Runs. It will try to match channel names in M3U files for failover and what not. It should be very flexible. No idea when I can get an alpha out to play with but I am pecking away!

**Comment by aniel300** [comment]
am  relative new to xteve and iptv in general, but how would the channel failover work ?  (it sounds like a nice thing to have since this way is just one channel showing but it will switch/failover if it goes down (all in the background)?

**Comment by kjetilhp** [comment]
a little late to the party, is the switch to c# and react a personal preference or based on technical issues with go? too bad about the choice of react over vue though ;) anyway, will you share the new repo before the alpha? WIll the failover work with all these ppv channels that changes names and appears dynamically on a daily basis?

**Comment by xtinct101** [comment]
All this sounds amazing, can't wait to try it out.

**Comment by SenexCrenshaw** [comment]
Plugging away here when I have free time. I am working out how to handle multiple EPGs/Channels and the UI. I have taken this time to learn React so this is taking longer than I thought. I would very much like a beta out this year and am trying my best to make that happen.

**Comment by SenexCrenshaw** [comment]
> am relative new to xteve and iptv in general, but how would the channel failover work ? (it sounds like a nice thing to have since this way is just one channel showing but it will switch/failover if it goes down (all in the background)?

It would need to be proxied/buffered through the server so when a backend stream fails it will use the other stream. I plan on adding in statistics for this as well so we know how the channels behave.

**Comment by SenexCrenshaw** [comment]
> a little late to the party, is the switch to c# and react a personal preference or based on technical issues with go? too bad about the choice of react over vue though ;) anyway, will you share the new repo before the alpha? WIll the failover work with all these ppv channels that changes names and appears dynamically on a daily basis?

With out starting a serious debate... C#  is a personal preference since its so easy and portable now-a-days. VS Code/Studio are solid IDEs. 

I am a huge VUE fan as well. I decided to go React since a lot of people out there know React and it was another framework for me to learn. I have been happy with react/redux so far. You could re-write the UI in anything you would like. The server is c# asp.net core and the frontend can be anything.

Once I get the HDHRs worked out, the buffer mechanism and cleanup I will share. 

"WIll the failover work with all these ppv channels that changes names and appears dynamically on a daily basis?" . You can create channels that will use those streams. Like you said though they are random... Maybe a feature where we can do regex matching for streams? Dunno. I would have to look into that. The bigger questions is the EPG for these.....

**Comment by kjetilhp** [comment]
> You could re-write the UI in anything you would like. The server is c# asp.net core and the frontend can be anything.

Great! Looking forward to your first release

**Comment by deku-m** [comment]
Is there also a docker version coming :)?

Good luck on the rewrite of the app. Looking forward to it.

**Comment by SenexCrenshaw** [comment]
Another piece that I was avoiding was any kind of dB for the backend. I am using flat jsons files for storage which added a lot of time in developing. After all of that I still might just go with sqlite.


I plan to only publish docker images for releases.

I use openapi v3 as well so creating clients for the api is very easy.

I'll get back to it and send an update in a week or so.

**Comment by aniel300** [comment]
ok am waiting on this

**Comment by GwynethLlewelyn** [comment]
C#? _\*covers\*_

Hmm, such a pity that it is your personal preference... especially because your Go fork worked so flawlessly for me! Thanks for doing that, BTW.

Moving on...

**Comment by hvmzx** [comment]
Would love to know how it's been going. Also, wouldn't it be nice to have a discord or some kind of community chat where we can discuss ideas if you want to totally redo xteve and basically create a new project ?

**Comment by SenexCrenshaw** [comment]
> C#? _\*covers\*_
> 
> Hmm, such a pity that it is your personal preference... especially because your Go fork worked so flawlessly for me! Thanks for doing that, BTW.
> 
> Moving on...

I hear ya!

**Comment by SenexCrenshaw** [comment]
> Would love to know how it's been going. Also, wouldn't it be nice to have a discord or some kind of community chat where we can discuss ideas if you want to totally redo xteve and basically create a new project ? 



I like this idea. I have gotten a little backed up with life but will try to get something out soon. I am treating this first iteration as a poc and will share when things work at the basic level. It's close. Cool thing you don't even need a UI. Everything can be done with the command line or whatever client you like to code in.

**Comment by SenexCrenshaw** [comment]
My goal is a poc by end of the first week in January. Assume festures will be missing. My goals for the poc are:

Multiples M3U files
Multiple EPG files
Icon caching
HDHRs will have a collection of channels and associated EPGs/programmes and tuner count
Buffering in code not using external programs
Statistics around channel usage, Dropped connections, Fail over, Usage... 


After that we all will have to start a feature list....

**Comment by GwynethLlewelyn** [comment]
> After that we all will have to start a feature list....

Does GitHub have polls? 🤣

**Comment by lars18th** [comment]
Hi @SenexCrenshaw ,

What will happens with the xTeVe when you will start with the next project? Any idea of which xTeVe _fork_ will be the best mantained? I put this question because I want to use xTeVe **only** for managing my M3U playlists. So I'm not interested on the Plex/Emby support nor the HDHR emulation.

**Comment by SenexCrenshaw** [comment]
Not sure of the forks out there. I'll patch the buffer bug in my existing one soon.

**Comment by lars18th** [comment]
> Not sure of the forks out there. I'll patch the buffer bug in my existing one soon.

Then I'll continue supporting your xTeVe almost for simple patches? And you'll accept PRs from other users?

**Comment by gdomod** [comment]
if i had a wish, i would get a failover / fallback stream
and a regex mapping config on import m3u list
m3u -> regex like ~das.*erste(.?HD) -> channel group "Das Erste HD" (list1:channel2, list2:channel4 etc)
channel group are mapped to xmltv

**Comment by hvmzx** [comment]
It would be nice to create groups based on the title of the stream. My m3us do not have any groups but they have prefixes like [AR] or :US:. If based on that prefix we can create groups, it could help tons !

**Comment by SenexCrenshaw** [comment]
Great suggestions. 
I was hoping to get a POC out this week, but it is just impossible. I haven't written a line a code in weeks. You know how it is... life and what nots. 

My goal is to start up again next week. In the meantime, we can discuss these features.  I want the new program to handle more than an HDHR emulator. 

I believe a channel is a collection of M3U streams, an icon and an EPG. The collection of streams would be used for failover, but do we ever envision a channel that switches streams after a program? Maybe even with a custom EPG.... 

I want other IPTV programs to use it including the front-end UI.

@lars18th I use xTeVe for the same thing and want more flexibility.

**Comment by lars18th** [comment]
Hi @SenexCrenshaw ,

Perfect if you wish to discuss some _functionalities_ before starting to code a new project!

From my point of view the suggestions for a _new_ project are:
- Remove the HDHR emulation functionality (almost as a central point). My objective is to use the tool as a _MANAGER_ of M3U playlists. In my case, the playlists (in plural) will come from SAT>IP and HDHR tuners. So I want to centralize the: channel scanning, the channel ordering, the metadata editing (picons, xmltv ids, channel ordering, etc.), EPG management (xmltv sources), automatic update of the playlists, etc.
- Daemon functionality with native multiplatform support (Linux/Windows/Android plus x86/ARM architectures) and with docker support.
- CLI and HTTP-GUI support.
- Support for plugins and/or scripting.

These are all of my wishes. But I'm sure that they are complex. What is your point of view?

**Comment by SenexCrenshaw** [comment]
Its a good list. I was using xteve for basically the same reasons you are.  
#1 I agree with  except the proxing part. Ill still add that
#2 Its .net core which runs on those arcs, dont forget macs. Docker is what i develop for
#3 yup. i use openapi V3 so you can make a TON of clients for the API
#4 Something i would look into l8r.

I have actually put some time into this over the weekend. Its not very complicated to do but I took this opty to learn a few new frameworks so it has slowed me a little. 

Thanks for the feedback!

**Comment by lars18th** [comment]
Hi @SenexCrenshaw ,

I hope you'll open soon the repo of the new project. And regarding the project name: do you have considered one? Perhaps **xTeVe++**?

**Comment by Soldize** [comment]
HI @SenexCrenshaw 

I can't wait for your new repo to be released.

best reagards

**Comment by hvmzx** [comment]
The part where you can manage your EPG is so important, Xteve is pretty clunky and lacks a lot of features in that department.

**Comment by adriel** [comment]
The biggest thing for me would be to have tvg-chno support. Meaning, if the source m3u8 file has them, to auto assign them in Mapping (the "Ch. No." col), if there are no conflicts. I've spent so much time manually entering them.

**Comment by hvmzx** [comment]
> The biggest thing for me would be to have tvg-chno support. Meaning, if the source m3u8 file has them, to auto assign them in Mapping (the "Ch. No." col), if there are no conflicts. I've spent so much time manually entering them.

This definitely and a tvg id auto assign based on a match from the channel name and a tvg id database. 

You could push this to other things : Auto assign tvg id/chno and channel icon

**Comment by SenexCrenshaw** [comment]
Good news is I have spent a bit of time on this. Send me some examples of tvg id/chno if you can.

**Comment by SenexCrenshaw** [comment]
Sneak peek
![image](https://user-images.githubusercontent.com/35600301/212136792-0fcec389-0d01-45d3-b901-2b8f7892f512.png)

**Comment by xtinct101** [comment]
This looks very promising.
Will initial release include failover support?

**Comment by SenexCrenshaw** [comment]
> This looks very promising. Will initial release include failover support?

Thats the plan

**Comment by aniel300** [comment]
is there a roadmap of already implemented features and the ones coming later on ?

**Comment by hvmzx** [comment]
> Sneak peek
> 
> ![image](https://user-images.githubusercontent.com/35600301/212136792-0fcec389-0d01-45d3-b901-2b8f7892f512.png)
> 
> 

That looks sexy as hell, looking forward to using it !

My provider gives me just a list of channels, not many have tvg id and there are no channel icons unfortunately. This is going to be good !

**Comment by adriel** [comment]
> Good news is I have spent a bit of time on this. Send me some examples of tvg id/chno if you can.

https://i.mjh.nz/au/Sydney/raw-tv.m3u8

They don't all have a tvg-chno, but most do.

**Comment by SenexCrenshaw** [comment]
EPG editor. 
![image](https://user-images.githubusercontent.com/35600301/212233953-b60e144b-7471-4db4-bd6a-223f3b6b97de.png)

**Comment by hvmzx** [comment]
> EPG editor. 
> 
> ![image](https://user-images.githubusercontent.com/35600301/212233953-b60e144b-7471-4db4-bd6a-223f3b6b97de.png)
> 
> 

Will xTeve have this interface also ? Or is this only for the editor? 

I think it would be go a good idea to create a discord so we can easily and quickly exchange ideas and maybe test this out while give feedback

**Comment by lars18th** [comment]
Hi,

> I think it would be go a good idea to create a discord so we can easily and quickly exchange ideas and maybe test this out while give feedback

I think it will be preferable to share a preview in github. For example, @SenexCrenshaw could create the project, and publish a draft of the binary (if hi doesn't wants to publish the source initially). Then we could test it and provide feedback/ideas in a discussion forum in github (please remember that github projects have a section, disabled by default, for discussions).

I hope that we will have soon something to discuss about it.
Thank you @SenexCrenshaw !! 👍

**Comment by SenexCrenshaw** [comment]
I have a few items left to do:
Categories - almost done, dynamic categories is part of this. 
Multi-stream channels- I want this to be flexible where we can do failover, load balancing...
tvg-chno
Channel Groups - similar to HDHR emulation
Docker

After that ill post the code and we can start a discord from there. Lets hope i can find the time and get something out my EOM.

The backend is c# core with an api that works over https/ws(signalr). It is super fast and is easily added to. I am trying to make it as easy to read and work in. The API supports openapi v3 so you can create any client you want using code generators, including a cli.
The frontend is react which is working really nicely with the api/ws setup.

**Comment by lars18th** [comment]
Hi @SenexCrenshaw ,

> After that ill post the code and we can start a discord from there. 

Great! However, _please_ don't create any discord to discuss about the new project (or open it _optionally_ only. I have no problem with this particular service). I suggest to enable the *Discussions* section in the github project. Search for it at: Settings, General: scroll down until Features. It's disabled by default. I love it for other projects in which I'm participating.

Regards.

**Comment by hvmzx** [comment]
> I have a few items left to do:

Can't wait! There is a discord for Xteve already and it truly allows people to discuss more easily compared to GitHub discussions that feels less dynamic. As for the version you're working on, can't wait to test it out and give you feedback !

**Comment by aniel300** [comment]
can't wait for the docker version, will this have restreaming feature ?

**Comment by adriel** [comment]
> > I have a few items left to do:
> 
> Can't wait! There is a discord for Xteve already and it truly allows people to discuss more easily compared to GitHub discussions that feels less dynamic. As for the version you're working on, can't wait to test it out and give you feedback !

Personally not a fan of Discord, I find it silos and splits the info.
If you search for x error code, you may not get any hits on the GitHub project, then you have to sign up and invite yourself to the Discord and do a search there (which I find isn't as good as in GitHub).

But I have seen the xteve Discord and a fair few people seem to use it, so maybe it works better then I fear.

Just my 2c.

**Comment by SenexCrenshaw** [comment]
> can't wait for the docker version, will this have restreaming feature ?

Do you mean a proxy in the middle like xteve Buffering. That's a yes if so

**Comment by aniel300** [comment]
yes, something like that. can't wait to test it out

**Comment by SenexCrenshaw** [comment]
How about a name for this app? BTW this is really nothing to do with xteve anymore...

According to chatgpt it is 
StreamMaster

Which i love!

**Comment by Soldize** [comment]
Will there be predefined options for Hardware accelerated video transcoding for ffmpeg?? For example, vaapi, Nvidia NVDEC & NVENC, Intel QuickSync Video, AMD AMF and so on? 

best regards

**Comment by Soldize** [comment]
and is it possible to set my own Wallpaper for "no more Streams available" ??

**Comment by lars18th** [comment]
> According to chatgpt it is StreamMaster

**StreamMaster** sounds good! 👍 

> Will there be predefined options for Hardware accelerated video transcoding for ffmpeg??

If the project will support some _Stream Processing_ then you could use FFmpeg or any other tool to achieve it. And then you can use accelerated transcoding if you need it and if your hardware supports it (it's a mere ffmpeg parameter configuration). However, I feel this could be considered as a manual user configuration. I feel it will be more positive in focusing on the _Channel List Administration_ functionality. Do you think so?

**Comment by Soldize** [comment]
What I would also like to see is multiline support.

Scenario: I have 2 lines from a provider with 2 connections each. If the 2 connections from the 1st line are in use then the system takes a connection from the 2nd line.

This function would be phenomenal (in Germany we would say "Ober Hammer Affen Titten Geil) and would have no other Tv Proxy <3

**Comment by aniel300** [comment]
isnt that what "xteve Buffering" is for ?

**Comment by Soldize** [comment]
No Xteve has no multiline support! You can add multiple M3u lists but that's all.

**Comment by aniel300** [comment]
yes but my point is that the buffering acts as a restreaming server doesnt it ?

**Comment by Soldize** [comment]
yes like restreaming

**Comment by aniel300** [comment]
so yes if am not mistaking that is what xteve buffering is supposed to do

**Comment by SenexCrenshaw** [comment]
Little tease
Dynamic Channel Categories. 

<img width="1265" alt="image" src="https://user-images.githubusercontent.com/35600301/213018026-85f95ad0-3980-4c8c-9720-6289c39584b8.png">

**Comment by SenexCrenshaw** [comment]
> Will there be predefined options for Hardware accelerated video transcoding for ffmpeg?? For example, vaapi, Nvidia NVDEC & NVENC, Intel QuickSync Video, AMD AMF and so on?
> 
> best regards

yet to be determined. I am hoping to not use an external program.....

**Comment by SenexCrenshaw** [comment]
> and is it possible to set my own Wallpaper for "no more Streams available" ??

Yep. icons can be set to anything you want. There is a default setting for channel icons that have no/bad URLs.

**Comment by SenexCrenshaw** [comment]
> What I would also like to see is multiline support.
> 
> Scenario: I have 2 lines from a provider with 2 connections each. If the 2 connections from the 1st line are in use then the system takes a connection from the 2nd line.
> 
> This function would be phenomenal (in Germany we would say "Ober Hammer Affen Titten Geil) and would have no other Tv Proxy <3

Yep. working this out. I expect to be able to add streams to channels and it will use that list and keep track of the streams health.

**Comment by Soldize** [comment]
> > What I would also like to see is multiline support.
> > Scenario: I have 2 lines from a provider with 2 connections each. If the 2 connections from the 1st line are in use then the system takes a connection from the 2nd line.
> > This function would be phenomenal (in Germany we would say "Ober Hammer Affen Titten Geil) and would have no other Tv Proxy <3
> 
> Yep. working this out. I expect to be able to add streams to channels and it will use that list and keep track of the streams health.

Shut the front door and take my money <3 <3 <3

**Comment by aniel300** [comment]
when are we getting the first release ? can't wait to test it out.

**Comment by aniel300** [comment]
does channel tag/grouping works with emby with this new version u working on ?

**Comment by deku-m** [comment]
Very nice work, looking forward to it @SenexCrenshaw 

Isnt it better to start a discussion page rather then in a support ticket. Maybe on the new repo?

**Comment by lars18th** [comment]
> Isnt it better to start a discussion page rather then in a support ticket. Maybe on the new repo?

Yes. I've recommeded this before: open the project, enable discussion board, and not commit any code until the first release.

**Comment by SenexCrenshaw** [comment]
Moving over to the new repo. I will post code when its ready to be tested. Discussions have been enabled:
[StreamMaster](https://github.com/SenexCrenshaw/StreamMaster)
[Discussions](https://github.com/SenexCrenshaw/StreamMaster/discussions)
